<a href="https://colab.research.google.com/github/shadirah/OA-COIL/blob/main/OA_Coil_Program_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
#mount drive
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


# **TASK 1: IMPORT DATA**

In [2]:
import pandas as pd

try:
    # Read the CSV file into a pandas DataFrame
    df = pd.read_excel('/content/drive/MyDrive/OA Coil Program/EUR_CHF Historical Data-1.xlsx')

    # Print the column names
    print(df.columns)
except FileNotFoundError:
    print("Error: The CSV file was not found at the specified path.")
    print("Please check if the file exists at '/content/drive/MyDrive/OA Coil Program/EUR_CHF Historical Data-1.xlsx' and verify the path is correct.")
    df = None # Assign None to df to avoid NameError if further code tries to use it

Index(['Date', 'Price', 'European Union GDP Growth',
       'Switzerland GDP Growth', 'EUR Gold Prices (onz)',
       'CHF Gold Prices (onz)', 'Oil Prices (Mean)'],
      dtype='object')


# **TASK 2: CALCULATE DAILY RETURN**

In [3]:
# Calculate the daily returns of 'Price'
df['forex_returns'] = df['Price'].pct_change() * 100

# Drop rows with NaN values in the 'forex_returns' column
df.dropna(subset=['forex_returns'], inplace=True)

# Display the updated DataFrame with the new 'forex_returns' column
display(df.head())

Date   Price  European Union GDP Growth  Switzerland GDP Growth  \
1 2024-06-27  0.9616                      0.003                   0.004   
2 2024-06-26  0.9581                      0.003                   0.004   
3 2024-06-25  0.9584                      0.003                   0.004   
4 2024-06-24  0.9580                      0.003                   0.004   
5 2024-06-21  0.9555                      0.003                   0.004   

   EUR Gold Prices (onz)  CHF Gold Prices (onz)  Oil Prices (Mean)  \
1                 2161.7                 2078.2          28.528027   
2                 2161.7                 2078.2          28.528027   
3                 2161.7                 2078.2          28.528027   
4                 2161.7                 2078.2          28.528027   
5                 2161.7                 2078.2          28.528027   

   forex_returns  
1      -0.124636  
2      -0.363977  
3       0.031312  
4      -0.041736  
5      -0.260960

# **TASK 3: DEFINE THE FEATURES (X) AND THE TARGET VARIABLE (y)**


In [4]:
# Clean and convert relevant columns to numeric
for col in ['European Union GDP Growth', 'Switzerland GDP Growth']:
    if col in df.columns:
        df[col] = df[col].astype(str).str.replace('%', '', regex=False).astype(float) / 100

for col in ['EUR Gold Prices (onz)', 'CHF Gold Prices (onz)']:
    if col in df.columns:
        df[col] = df[col].astype(str).str.replace(',', '', regex=False).astype(float)

# Define the features (X) and target variable (y)
X = df[['European Union GDP Growth', 'Switzerland GDP Growth', 'EUR Gold Prices (onz)', 'CHF Gold Prices (onz)', 'Oil Prices (Mean)']]
y = df['forex_returns']

# Display the first few rows of X and y
display(X.head())
display(y.head())

European Union GDP Growth  Switzerland GDP Growth  EUR Gold Prices (onz)  \
1                    0.00003                 0.00004                 2161.7   
2                    0.00003                 0.00004                 2161.7   
3                    0.00003                 0.00004                 2161.7   
4                    0.00003                 0.00004                 2161.7   
5                    0.00003                 0.00004                 2161.7   

   CHF Gold Prices (onz)  Oil Prices (Mean)  
1                 2078.2          28.528027  
2                 2078.2          28.528027  
3                 2078.2          28.528027  
4                 2078.2          28.528027  
5                 2078.2          28.528027

1   -0.124636
2   -0.363977
3    0.031312
4   -0.041736
5   -0.260960
Name: forex_returns, dtype: float64

# **TASK 4: SPLIT THE DATA**

**Explanation**:
Split the data into training and testing sets using an 80/20 ratio.



In [5]:
from sklearn.model_selection import train_test_split

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Display the shapes of the resulting sets to verify the split
print("Shape of X_train:", X_train.shape)
print("Shape of X_test:", X_test.shape)
print("Shape of y_train:", y_train.shape)
print("Shape of y_test:", y_test.shape)

Shape of X_train: (728, 5)
Shape of X_test: (182, 5)
Shape of y_train: (728,)
Shape of y_test: (182,)


# **TASK 5: TRAIN THE MODEL**



**Explanation**:
Train a Random Forest Regressor model using the training data.



In [6]:
from sklearn.ensemble import RandomForestRegressor

# Instantiate a RandomForestRegressor object
model = RandomForestRegressor(random_state=42)

# Train the model using the X_train and y_train dataframes
model.fit(X_train, y_train)

RandomForestRegressor(random_state=42)

## **TASK 6: EVALUATE THE MODEL**



**Explanation**:
Make predictions on the test data and evaluate the model's performance using appropriate metrics.



In [7]:
from sklearn.metrics import mean_squared_error, r2_score

# Make predictions on the test set
y_pred = model.predict(X_test)

# Calculate Mean Squared Error (MSE)
mse = mean_squared_error(y_test, y_pred)

# Calculate R-squared score
r2 = r2_score(y_test, y_pred)

# Print the calculated metrics
print(f"Mean Squared Error (MSE): {mse}")
print(f"R-squared (R2) Score: {r2}")

Mean Squared Error (MSE): 0.13388846850802555
R-squared (R2) Score: -0.16318129227678568


# **TASK 7: PREDICTIVE MODEL**

In [8]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.tree import DecisionTreeRegressor
from xgboost import XGBRegressor
!pip install catboost
from catboost import CatBoostRegressor

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42)

models = {
    "Decision Tree": DecisionTreeRegressor(random_state=42),
    "Random Forest": RandomForestRegressor(random_state=42),
    "Gradient Boosting": GradientBoostingRegressor(random_state=42),
    "XGBoost": XGBRegressor(random_state=42),
    "CatBoost": CatBoostRegressor(random_state=42, verbose=0)
}

results = {}
feature_importances = {}

for name, model in models.items():
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)

    results[name] = {
        "MSE": mean_squared_error(y_test, y_pred),
        "R2": r2_score(y_test, y_pred)
    }

    if name == "CatBoost":
        fi = model.get_feature_importance()
    else:
        fi = model.feature_importances_

    feature_importances[name] = pd.Series(fi, index=X.columns)

results_df = pd.DataFrame(results).T
print("Model Performance:\n", results_df)

Model Performance:
                         MSE        R2
Decision Tree      0.133235 -0.157501
Random Forest      0.133888 -0.163181
Gradient Boosting  0.131766 -0.144744
XGBoost            0.133195 -0.157153
CatBoost           0.133221 -0.157386


## Summary of Tree-Based Model Performance and Feature Importances

This summary compares the performance metrics (Mean Squared Error (MSE) and R-squared (R2) Score) and feature importances for the Random Forest Regressor, Gradient Boosting Regressor, XGBoost Regressor, CatBoost Regressor, and Decision Tree Regressor models, all trained on the same features and target variable ('forex_returns').

### Performance Comparison (Lower MSE is better, Higher R2 is better):

| Model                       | Mean Squared Error (MSE) | R-squared (R2) Score |
|:----------------------------|:-------------------------|:---------------------|
| Random Forest Regressor     | 0.133888                 | -0.163181            |
| Gradient Boosting Regressor | 0.131766                 | -0.144744            |
| XGBoost Regressor           | 0.133195                 | -0.157153            |
| CatBoost Regressor          | 0.133221                 | -0.157386            |
| Decision Tree Regressor     | 0.133235                 | -0.157501            |

**Observations:**
*   All models exhibit similar MSE values, hovering around 0.133. This suggests that none of the models are significantly outperforming the others in terms of absolute error.
*   The R-squared scores for all models are negative, ranging from approximately -0.145 to -0.163. A negative R2 score indicates that the models perform worse than a simple horizontal line (the mean of the actual values). This suggests that the chosen features might not be highly predictive of 'forex_returns' or that the relationship is highly non-linear and not well captured by these models without further tuning or feature engineering.
*   Among the models, the **Gradient Boosting Regressor** shows a slightly lower MSE and a slightly less negative R2 score, indicating marginally better performance in this specific evaluation.

### Feature Importance Comparison:

**Random Forest Regressor Feature Importances:**
Oil Prices (Mean)            0.347415
CHF Gold Prices (onz)        0.277093
EUR Gold Prices (onz)        0.208443
Switzerland GDP Growth       0.096446
European Union GDP Growth    0.070603

**Gradient Boosting Machine Feature Importances:**
Oil Prices (Mean)            0.383559
CHF Gold Prices (onz)        0.275767
EUR Gold Prices (onz)        0.148144
Switzerland GDP Growth       0.116983
European Union GDP Growth    0.075548

**XGBoost Regressor Feature Importances:**
Oil Prices (Mean)            0.345511
Switzerland GDP Growth       0.230633
CHF Gold Prices (onz)        0.165305
EUR Gold Prices (onz)        0.161622
European Union GDP Growth    0.096928

**CatBoost Regressor Feature Importances:**
Oil Prices (Mean)            26.683171
EUR Gold Prices (onz)        20.462624
CHF Gold Prices (onz)        20.037275
Switzerland GDP Growth       18.316533
European Union GDP Growth    14.500397

**Decision Tree Regressor Feature Importances:**
Oil Prices (Mean)            0.385425
CHF Gold Prices (onz)        0.260973
EUR Gold Prices (onz)        0.203389
Switzerland GDP Growth       0.117056
European Union GDP Growth    0.033157

**Key Takeaways from Feature Importances:**
*   Across all tree-based models, **'Oil Prices (Mean)'** consistently emerges as the most important feature for predicting 'forex_returns'. This highlights its significant influence on the target variable.
*   **'CHF Gold Prices (onz)'** and **'EUR Gold Prices (onz)'** are also consistently ranked as highly important features, often second or third.
*   **'Switzerland GDP Growth'** and **'European Union GDP Growth'** generally show lower importance compared to 'Oil Prices' and 'Gold Prices', although their relative importance can vary slightly between models (e.g., Switzerland GDP Growth is more important than gold prices in XGBoost but less in others).

### Conclusion:
While all tree-based models show similar (and poor) predictive performance on this dataset (indicated by negative R2 scores), their feature importance analyses consistently point to **'Oil Prices (Mean)'** as the primary driver, followed by gold prices. This suggests that for future model improvements, focusing on these highly influential features, potentially through more sophisticated feature engineering or exploring different time-series models, would be beneficial. The consistently negative R2 scores also suggest that the current set of features and models might not be adequately capturing the underlying patterns in 'forex_returns'.

# **TASK 8: FEATURE IMPORTANCE**

In [9]:
print("\nFeature Importances for All Models:")
for name, fi in feature_importances.items():
    print(f"\n{name}:")
    print(fi.sort_values(ascending=False))


Feature Importances for All Models:

Decision Tree:
Oil Prices (Mean)            0.385425
CHF Gold Prices (onz)        0.260973
EUR Gold Prices (onz)        0.203389
Switzerland GDP Growth       0.117056
European Union GDP Growth    0.033157
dtype: float64

Random Forest:
Oil Prices (Mean)            0.347415
CHF Gold Prices (onz)        0.277093
EUR Gold Prices (onz)        0.208443
Switzerland GDP Growth       0.096446
European Union GDP Growth    0.070603
dtype: float64

Gradient Boosting:
Oil Prices (Mean)            0.383559
CHF Gold Prices (onz)        0.275767
EUR Gold Prices (onz)        0.148144
Switzerland GDP Growth       0.116983
European Union GDP Growth    0.075548
dtype: float64

XGBoost:
Oil Prices (Mean)            0.345511
Switzerland GDP Growth       0.230633
CHF Gold Prices (onz)        0.165305
EUR Gold Prices (onz)        0.161622
European Union GDP Growth    0.096928
dtype: float32

CatBoost:
Oil Prices (Mean)            26.683171
EUR Gold Prices (onz)        20

# **TASK 9: OPTIONAL EXTENSIONS (LASSO, PCA &OLS)**

##LASSO

In [10]:
from sklearn.linear_model import Lasso

# Define the features (X) and target variable (y) using available columns
X = df[['European Union GDP Growth', 'Switzerland GDP Growth', 'EUR Gold Prices (onz)', 'CHF Gold Prices (onz)', 'Oil Prices (Mean)']]
y = df['forex_returns']

# Instantiate a Lasso model without an intercept
lasso_model = Lasso(alpha=1.0, fit_intercept=False) # You might need to tune the alpha parameter

# Fit the model
lasso_model.fit(X, y)

# Print the coefficients
print("Lasso Coefficients:")
for feature, coef in zip(X.columns, lasso_model.coef_):
    print(f"{feature}: {coef}")

Lasso Coefficients:
European Union GDP Growth: 0.0
Switzerland GDP Growth: 0.0
EUR Gold Prices (onz): 0.0
CHF Gold Prices (onz): 6.820099320223385e-06
Oil Prices (Mean): 0.0


## PRINCIPAL COMPENENT ANALYSIS (PCA)

In [11]:
from sklearn.decomposition import PCA
import numpy as np

# Define the features (X)
X = df[['European Union GDP Growth', 'Switzerland GDP Growth', 'EUR Gold Prices (onz)', 'CHF Gold Prices (onz)', 'Oil Prices (Mean)']]

# Instantiate PCA with the number of components equal to the number of features
pca = PCA(n_components=X.shape[1])

# Fit PCA on the features
pca.fit(X)

# Get the loading factors (components)
loading_factors = pca.components_

# Create a DataFrame for better visualization
loading_factors_df = pd.DataFrame(loading_factors, columns=X.columns, index=[f'PC{i+1}' for i in range(pca.n_components_)])

# Print the loading factors
print("Principal Component Loading Factors:")
display(loading_factors_df)

Principal Component Loading Factors:


European Union GDP Growth  Switzerland GDP Growth  EUR Gold Prices (onz)  \
PC1              -1.569097e-07           -1.515010e-07           8.289452e-01   
PC2              -1.496525e-07           -2.213972e-07           5.479885e-02   
PC3               1.003024e-06            9.349011e-07          -5.566389e-01   
PC4               4.267209e-01            9.043834e-01           8.924440e-07   
PC5               9.043834e-01           -4.267209e-01           3.491527e-07   

     CHF Gold Prices (onz)  Oil Prices (Mean)  
PC1           5.554547e-01      -6.572565e-02  
PC2           3.629150e-02       9.978377e-01  
PC3           8.307545e-01       3.546141e-04  
PC4          -9.351014e-07       2.496590e-07  
PC5          -3.777731e-07       3.552198e-08

## OLS Regression

In [12]:
import statsmodels.api as sm

# Add a constant (intercept) to the features
X_with_intercept = sm.add_constant(X)

# Create the OLS model
model_ols = sm.OLS(y, X_with_intercept)

# Fit the model with robust standard errors (HC1)
results_ols = model_ols.fit(cov_type='HC1')

# Print the regression summary
print(results_ols.summary())

                            OLS Regression Results                            
Dep. Variable:          forex_returns   R-squared:                       0.001
Model:                            OLS   Adj. R-squared:                 -0.004
Method:                 Least Squares   F-statistic:                    0.2566
Date:                Mon, 01 Dec 2025   Prob (F-statistic):              0.937
Time:                        13:50:16   Log-Likelihood:                -326.66
No. Observations:                 910   AIC:                             665.3
Df Residuals:                     904   BIC:                             694.2
Df Model:                           5                                         
Covariance Type:                  HC1                                         
                                coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------------------
const                 